python/ray/data/_internal/progress_bar.py

```diff
@@ -83,6 +91,34 @@ def __init__(
                 needs_warning = False
             self._bar = None
 
+    def _truncate_name(self, name: str) -> str:
+        ctx = ray.data.context.DataContext.get_current()
+        if (
+            not ctx.enable_progress_bar_name_truncation
+            or len(name) <= self.MAX_NAME_LENGTH
+        ):
+            return name
+
+        if log_once("ray_data_truncate_operator_name"):
+            logger.warning(
+                f"Truncating long operator name to {self.MAX_NAME_LENGTH} characters."
+                "To disable this behavior, set `ray.data.DataContext.get_current()."
+                "DEFAULT_ENABLE_PROGRESS_BAR_NAME_TRUNCATION = False`."
+            )
+        op_names = name.split("->")
+        # Include as many operators as possible without exceeding `MAX_NAME_LENGTH`.
+        # Always include the first and last operator names so
+        # it is easy to identify the DAG.
+        truncated_op_names = [op_names[0]]
+        for i, op_name in enumerate(op_names[1:-1]):
+            if len("->".join(truncated_op_names)) + len(op_name) > self.MAX_NAME_LENGTH:
+                truncated_op_names.append("...")
+                break
+            truncated_op_names.append(op_name)
+        if len(op_names) > 1:
+            truncated_op_names.append(op_names[-1])
+        return "->".join(truncated_op_names)
```

Comment:
Nit: Not a big deal, but I think there are some edge cases where the truncated name can exceed `MAX_NAME_LENGTH` because we don't account for the last name or the additional "->"s.

```suggestion
        op_names = name.split("->")
        if len(op_names) == 1:
            return op_names[0]
        else:
            # Include as many operators as possible without exceeding `MAX_NAME_LENGTH`.
            # Always include the first and last operator names so
            # it is easy to identify the DAG.
            truncated_op_names = [op_names[0]]
            for op_name in op_names[1:-1]:
                if len("->".join(truncated_op_names)) + len("->") + len(op_name) + len("->") + len(op_names[-1]) > self.MAX_NAME_LENGTH:
                    truncated_op_names.append("...")
                    break
                truncated_op_names.append(op_name)
            truncated_op_names.append(op_names[-1])
            return "->".join(truncated_op_names)
```